hackathon_site/hardware/test_api.py

```diff
@@ -194,3 +194,244 @@ def test_hardware_get_success(self):
         data = response.json()
 
         self.assertEqual(expected_response, data["results"])
+
+
+class HardwareListTestCase(SetupUserMixin, APITestCase):
+    def setUp(self):
+        super().setUp()
+        # Create 3ish hardware objects
+        self.hardware1 = Hardware.objects.create(
+            name="hardware1",
+            model_number="model",
+            manufacturer="manufacturer",
+            datasheet="/datasheet/location/",
+            notes="notes",
+            quantity_available=0,
+            max_per_team=1,
+            picture="/picture/location",
+        )
+
+        self.hardware2 = Hardware.objects.create(
+            name="arduino",
+            model_number="model",
+            manufacturer="manufacturer",
+            datasheet="/datasheet/location/",
+            notes="notes",
+            quantity_available=4,
+            max_per_team=1,
+            picture="/picture/location",
+        )
+
+        self.hardware3 = Hardware.objects.create(
+            name="hardware3",
+            model_number="model",
+            manufacturer="manufacturer",
+            datasheet="/datasheet/location/",
+            notes="notes",
+            quantity_available=4,
+            max_per_team=1,
+            picture="/picture/location",
+        )
+
+        self.view = reverse("api:hardware:hardware-list")
+
+    def _build_filter_url(self, **kwargs):
+        return (
+            self.view + "?" + "&".join([f"{key}={val}" for key, val in kwargs.items()])
+        )
+
+    def test_user_not_logged_in(self):
+        response = self.client.get(self.view)
+        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
+
+    def test_search_by_name(self):
+        self._login()
+
+        url = self._build_filter_url(search="arduino")
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+        data = response.json()
+        expected_data = {
+            # Your expected data
+            "id": 2,
+            "name": "arduino",
+            "model_number": "model",
+            "manufacturer": "manufacturer",
+            "datasheet": "/datasheet/location/",
+            "quantity_available": 4,
+            "notes": "notes",
+            "max_per_team": 1,
+            "picture": "http://testserver/media/picture/location",
+            "categories": [],
+            "quantity_remaining": 4,
+        }
+
+        self.assertEqual(data["results"][0], expected_data)
+
+    def test_in_stock_filter(self):
```

Comment:
So that you're actually sure this test verifies that `quantity_remaining` is 0 (not that `quantity_available` is 0), you should set the `quantity_available` on hardware1 to 1, then create an order item for it.

Also as in my other comment, no need to test the full response format, just what's been returned. This test could become:

```python
    def test_in_stock_true(self):
        self._login()

        OrderItem.objects.create(hardware=self.hardware1, order=self.order)

        url = self._build_filter_url(in_stock="true")
        response = self.client.get(url)
        self.assertEqual(response.status_code, status.HTTP_200_OK)

        data = response.json()
        results = data["results"]
        returned_ids = [res["id"] for res in results]
        self.assertEqual(returned_ids, [2, 3])
```

You'll also need to add this to the bottom of `setUp()` for that to work:

```python
self.team = Team.objects.create()
self.order = Order.objects.create(status="Submitted", team=self.team)
```

Same thing for when `in_stock` is false, but in a separate test method.

If you do that, the test will actually fail! The `quantity_remaining` stuff has now been merged in to develop, so you can merge in the develop branch and update the filter method to use it instead of `quantity_available`.